File: c/childs.c
#include <sys/wait.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>

int main(void)
{
        pid_t pids[10];
        int i;

        for (i = 9; i >= 0; --i) {
                pids[i] = fork();
                if (pids[i] == 0) {
                        printf("hello %i\n",i);
                        sleep(i+1);
                        _exit(0);
                }
                printf("child %d\n",pids[i]);
        }

        for (i = 9; i >= 0; --i) {
                printf("wait  %d\n",pids[i]);
                waitpid(pids[i], NULL, 0);
         }

        return 0;
}

File: perl/childs.pl
#!/usr/bin/perl

use strict;
use warnings;

my $pid = fork;

die "Cannot fork: $!" unless defined $pid;

if ($pid == 0) {
    print "Child start\n";
    my $end;
    local $SIG{HUP} = sub { $end = 1 };

    until ($end) {
        print "Sleep 1\n";
        sleep 1;
    }
    function2();
}
else {
    print "Parent start\n";
    sleep 5;
    kill HUP => $pid;
    waitpid($pid, 0);
}

sub function2 {
    print "END\n";
}
